syntax = "proto3";

option objc_class_prefix = "HLW";

// The ClientHandler service definition.
service ClientHandler {
  // Lists users
  rpc ListUsers (ListRequest) returns (ListReply) {}
  // Requests to Login
  rpc Login (LoginRequest) returns (LoginReply) {}
  // Requests to Logout
  rpc Logout (LogoutRequest) returns (LogoutReply) {}
  // Attempts to Send message
  rpc Send (SendRequest) returns (SendReply) {}
  // Attempts to receive new messages
  rpc GetMessages (GetRequest) returns (GetReply) {}
  // Attempts to delete account
  rpc Delete (DeleteRequest) returns (DeleteReply) {}
  // Creates backup chain
  rpc GetBackups (BackupRequest) returns (BackupReply) {}

  // Methods to be used by ServerWorkers
  rpc AddUser(LoginRequest) returns (Empty) {}
  rpc RemoveUser(DeleteRequest) returns (Empty) {}
  rpc SetUserStatus(SetStatusRequest) returns (Empty) {}
  rpc AddMessage(SendRequest) returns (Empty) {}
  rpc DeleteMessages(GetRequest) returns (Empty) {}

  rpc PullData(Empty) returns (Data) {}
  rpc CheckClock(Empty) returns (Clock) {}

}

message Clock {
  int64 clock = 1;
}

message User {
  string user = 1;
  bool online = 2;
}

message Data {
  Clock clock = 1;
  repeated User users = 2;
  repeated UnreadMessage messages = 3;
}



message Empty {
  
}

message SetStatusRequest {
  string user = 1;
  bool status = 2;
}

message ServerInfo {
  string host = 1;
  string port = 2;
}

message BackupRequest {
}

message BackupReply {
  int32 status = 1;
  string errormessage = 2;
  repeated ServerInfo serverinfo = 3;
}

message LoginRequest {
  string user = 1;
}

message LogoutRequest{
    string user = 1;
}

message ListRequest {
    string args = 1;
}

message SendRequest {
  string user = 1;
  string message = 2;
  string target = 3;
}

message DeleteRequest {
    string user = 1;
}

message GetRequest {
  string user = 1;
}

message LoginReply {
  int32 status = 1;
  string errormessage = 2;
  string user = 3;
}

message LogoutReply {
    int32 status = 1;
    string errormessage = 2;
    string user = 3;
}

message ListReply {
  int32 status = 1;
  string errormessage = 2;
  string wildcard = 3;
  repeated string user = 4;
}

message DeleteReply {
    int32 status = 1;
    string errormessage = 2;
    string user = 3;
}

message SendReply {
    int32 status = 1;
    string errormessage = 2;
    string user = 3;
    string message = 4;
    string target = 5;
}

message UnreadMessage {
  string sender = 1;
  string message = 2;
  string receiver = 3;
}

message GetReply {
  int32 status = 1;
  string errormessage = 2;
  repeated UnreadMessage message = 3;
}